### PR TITLE
Metalamp/nft marketplace/refactor auction

### PR DESCRIPTION
**Pull request by olgaklimenko**

### CODE REVIEW

File: MetaLamp/nft-marketplace/src/Plutus/Contracts/NftMarketplace/OnChain/Core/NFT.hs [hunk 1]
```diff
@@ -17,84 +17,41 @@
...
 
 -- Category = [BuiltinByteString]
 -- 1. acts as a list of category with nested subcategories
 -- 2. acts as a list of tags
 type IpfsCid = BuiltinByteString
 type IpfsCidHash = BuiltinByteString
--- type Auction = (ThreadToken, PubKeyHash, Value, POSIXTimeT)
 type Category = [BuiltinByteString]
-type LotLink = Either Sale.Sale Auction
+type LotLink = Either Sale.Sale Auction.Auction
```
Comment (performanceArtist) [inline MetaLamp/nft-marketplace/src/Plutus/Contracts/NftMarketplace/OnChain/Core/NFT.hs]: I think using `Either` to represent a custom type that just happens to have two alternatives is a bad practice. It is not consistent, extendable, and is just confusing overall. Either has a clear bias towards `Right` value, while any custom type should not have that, unless it is also a Monad, which is usually not the case. The only reason this is done is to reuse`either` function and `Either` constructors. But `either` can be easily swapped with `case` and an additional data constructor only costs us an import, potentially improving readability. 

I suggest replacing this type as well as `PutLotRedeemer` and `RemoveLotRedeemer` with custom data.

Comment (olgaklimenko) [reply MetaLamp/nft-marketplace/src/Plutus/Contracts/NftMarketplace/OnChain/Core/NFT.hs]: Fully agree with you. And it's a good time to fix that.

File: MetaLamp/nft-marketplace/test/Marketplace/Spec/Auction.hs [hunk 2]
```diff
@@ -116,8 +116,8 @@ tests =
         buyOnAuctionTraceB
     ]]
 
-auctionValue :: Marketplace.Auction -> Value
-auctionValue = Auction.apAsset . Marketplace.fromAuction
+auctionValue :: Auction.Auction -> Value
+auctionValue = Auction.aAsset
```
Comment (performanceArtist) [inline MetaLamp/nft-marketplace/test/Marketplace/Spec/Auction.hs]: Any point in this alias?

File: MetaLamp/nft-marketplace/src/Plutus/Contracts/NftMarketplace/OnChain/Core/NFT.hs [hunk 3]
```diff
@@ -42,16 +42,40 @@ import qualified PlutusTx.AssocMap                      as AssocMap
...
+  deriving stock (Haskell.Eq, Haskell.Show, Haskell.Generic)
+  deriving anyclass (J.ToJSON, J.FromJSON)
+
+PlutusTx.unstableMakeIsData ''LotLink
+
+PlutusTx.makeLift ''LotLink
+
+Lens.makeClassy_ ''LotLink
+
+getSaleFromLot :: LotLink -> Maybe Sale.Sale
```
Comment (performanceArtist) [inline MetaLamp/nft-marketplace/src/Plutus/Contracts/NftMarketplace/OnChain/Core/NFT.hs]: No need for these two functions(`getAuctionFromLot`, `getAuctionFromLot`) - we can use `^?` instead.

Comment (performanceArtist) [reply MetaLamp/nft-marketplace/src/Plutus/Contracts/NftMarketplace/OnChain/Core/NFT.hs]: On the other hand, I remember having some issues with lenses in validators - is this the reason these are written by hand?

File: MetaLamp/nft-marketplace/src/Plutus/Contracts/NftMarketplace/OffChain/ID.hs [hunk 4]
```diff
@@ -37,7 +36,7 @@ import           Prelude                                                (Semigro
 import qualified Prelude                                                as Haskell
 import qualified Schema
 import           Text.Printf                                            (printf)
-
+import Plutus.Contracts.NftMarketplace.OnChain.Core.ID (InternalId(..))
 -- type UserItemId = Either Core.IpfsCid [Core.IpfsCid]
```
Comment (performanceArtist) [inline MetaLamp/nft-marketplace/src/Plutus/Contracts/NftMarketplace/OffChain/ID.hs]: Leftover comment.

File: MetaLamp/nft-marketplace/src/Plutus/Contracts/NftMarketplace/OnChain/Core/NFT.hs (hunk 3 above)
Comment (olgaklimenko) [reply MetaLamp/nft-marketplace/src/Plutus/Contracts/NftMarketplace/OnChain/Core/NFT.hs]: Good point, that operator works)

**Review by performanceArtist — APPROVED** [review]

**Pull request merged**